lib/charms/opensearch/v0/opensearch_snapshots.py

```diff
@@ -0,0 +1,1197 @@
+            alt_hosts=self.charm.alt_hosts,
+        )
+        logger.info("Restore of snapshot '%s' response: %s", snapshot_id, restore_resp)
+
+        # this only serves as documentation and should always be true if no previous HTTP error
+        assert restore_resp["snapshot"] == snapshot_id
+
+        # sanity check on the restore success
+        recovery_resp: list[dict[str, str]] = self.opensearch.request(
+            "GET", "_cat/recovery?format=json"
+        )
+        snapshot_recoveries = [
+            recovery
+            for recovery in recovery_resp
+            if (
+                recovery["type"] == "snapshot"
+                and recovery["repository"] == repo_name
+                and recovery["snapshot"] == snapshot_id
+            )
+        ]
+        restored_indices = set(
+            [recovery["index"] for recovery in snapshot_recoveries if recovery["stage"] == "done"]
+        )
+        expected_indices = set(snapshot.get("indices", []))
+        return expected_indices - restored_indices
+
+    @retry(stop=stop_after_attempt(3), wait=wait_fixed(3), reraise=True)
+    def close_snapshot_indices_open_in_cluster(
+        self, snapshot: dict[str, Any]
+    ) -> Tuple[list[str] | None, dict[str, Any] | None]:
+        """Close the non-system indices included in a given snapshot."""
+        if not (indices_to_close := self._get_snapshot_indices_open_in_cluster(snapshot)):
+            logger.info("No indices to close.")
+            return None, None
+
+        logger.info("Attempting closing the indices: %s", indices_to_close)
+        response = self.opensearch.request("POST", f"{','.join(indices_to_close)}/_close")
+
+        # verify that the relevant indices are closed
+        if response["acknowledged"] and response["shards_acknowledged"]:
+            logger.info("Successfully closed all indices: %s.", indices_to_close)
+            return indices_to_close, None
+
+        indices_failed_to_close = {
+            index: payload
+            for index, payload in response["indices"].items()
+            if not payload["closed"]
+        }
+        closed_indices = [
+            index for index in indices_to_close if index not in indices_failed_to_close
+        ]
+
+        logger.error("Failed to close some indices: \n%s", indices_failed_to_close)
+        return closed_indices, indices_failed_to_close
+
+    @retry(stop=stop_after_attempt(3), wait=wait_fixed(3), reraise=True)
+    def _get_snapshot_indices_open_in_cluster(self, snapshot: dict[str, Any]) -> list[str]:
+        """Fetch the current open indices in the current cluster."""
+        current_indices = ClusterState.indices(self.opensearch)
+        return sorted(
+            [
+                index
+                for index in snapshot.get("indices", [])
+                if index in current_indices
+                and index not in SYSTEM_INDICES
+                and current_indices[index]["state"] == "open"
```

Comment:
There doesn't appear to be a `state` key in this object, I believe it should be `status`
```suggestion
                and current_indices[index]["status"] == "open"
```

Reply:
This is still looking for `state` which does not exist

Reply:
Done